src/spi.rs

```diff
@@ -1641,68 +1654,160 @@ fn spi_create_transaction(
+                let mut ret_trans: *mut spi_transaction_t = ptr::null_mut();
+                esp!(unsafe {
+                    spi_device_get_trans_result(handle, &mut ret_trans as *mut _, delay::BLOCK)
+                })?;
+                unsafe { &mut *ret_trans }
+            };
+
+            // Write transaction to stable memory location
+            *trans = transaction;
+            esp!(unsafe { spi_device_queue_trans(handle, trans as *mut _, delay::BLOCK) })?;
+        }
+
+        let queued_transactions = min(MAX_QUEUED_TRANSACTIONS, queue_size) - queue_iter.len();
+        for _ in 0..queued_transactions {
+            let mut ret_trans: *mut spi_transaction_t = ptr::null_mut();
+            esp!(unsafe {
+                spi_device_get_trans_result(handle, &mut ret_trans as *mut _, delay::BLOCK)
+            })?;
+        }
     }
+    Ok(())
 }
 
 async fn spi_transmit_async(
     handle: spi_device_handle_t,
-    transaction: &mut spi_transaction_t,
-    _keep_cs_active: bool,
+    transactions: impl Iterator<Item = spi_transaction_t>,
+    queue_size: usize,
 ) -> Result<(), EspError> {
-    // This unfortunately means that this implementation is incorrect for esp-idf < 4.4.
-    // The CS pin should be kept active through transactions.
-    #[cfg(not(esp_idf_version = "4.3"))]
-    if _keep_cs_active {
-        transaction.flags |= SPI_TRANS_CS_KEEP_ACTIVE
-    }
-
-    let notification = Notification::new();
-
-    transaction.user = ptr::addr_of!(notification) as *mut _;
-
-    match esp!(unsafe { spi_device_queue_trans(handle, transaction as *mut _, delay::NON_BLOCK) }) {
-        Err(e) if e.code() == ESP_ERR_TIMEOUT => unreachable!(),
-        other => other,
-    }?;
+    let notifications: [Notification; MAX_QUEUED_TRANSACTIONS] = [
+        Notification::new(),
+        Notification::new(),
+        Notification::new(),
+        Notification::new(),
+        Notification::new(),
+        Notification::new(),
+        Notification::new(),
+        Notification::new(),
+        Notification::new(),
+        Notification::new(),
+    ];
+
+    let mut transaction_queue = [spi_transaction_t::default(); MAX_QUEUED_TRANSACTIONS];
+    let mut queue_iter = transaction_queue.iter_mut().take(queue_size);
+
+    let queued_transactions = Cell::new(0); // Total successful spi_device_queue_trans
+    let completed_transactions = Cell::new(0); // Total successful spi_device_get_trans_result
 
     with_completion(
         async {
-            notification.wait().await;
+            for (index, transaction) in transactions.enumerate() {
+                let slot = queue_iter.next();
+                let trans = if let Some(slot) = slot {
+                    slot
+                } else {
+                    // If the queue is full, we wait for the first transaction in the queue and use it
+                    // for the next one.
+
+                    notifications[completed_transactions.get() % MAX_QUEUED_TRANSACTIONS]
```

Comment:
Shouldn't this be `% min(MAX_QUEUED_TRANSACTIONS, queue_size)`?

Reply:
Both work. I chose not to use `min` to make the code shorter.
The notification at index `i` of `notifications` doesn't need to match the transaction at index `i` of `transaction_queue`.